In [1]:
import sys; sys.version

'3.6.8 |Anaconda, Inc.| (default, Dec 29 2018, 19:04:46) \n[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]'

## Reinforcement

### R-5.1
Execute the experiment from Code Fragment 5.1 and compare the results on your system to those we report in Code Fragment 5.2.

In [2]:
import sys  # provides getsizeof function
data=[]
for k in range(27):          # NOTE: must fix choice of n
    a = len(data)            # number of elements  
    b = sys.getsizeof(data)  # actual size in bytes
    print('Length: {0:3d}; Size in bytes: {1:4d}'.format(a, b))
    data.append(None)        # increase length by one

Length:   0; Size in bytes:   64
Length:   1; Size in bytes:   96
Length:   2; Size in bytes:   96
Length:   3; Size in bytes:   96
Length:   4; Size in bytes:   96
Length:   5; Size in bytes:  128
Length:   6; Size in bytes:  128
Length:   7; Size in bytes:  128
Length:   8; Size in bytes:  128
Length:   9; Size in bytes:  192
Length:  10; Size in bytes:  192
Length:  11; Size in bytes:  192
Length:  12; Size in bytes:  192
Length:  13; Size in bytes:  192
Length:  14; Size in bytes:  192
Length:  15; Size in bytes:  192
Length:  16; Size in bytes:  192
Length:  17; Size in bytes:  264
Length:  18; Size in bytes:  264
Length:  19; Size in bytes:  264
Length:  20; Size in bytes:  264
Length:  21; Size in bytes:  264
Length:  22; Size in bytes:  264
Length:  23; Size in bytes:  264
Length:  24; Size in bytes:  264
Length:  25; Size in bytes:  264
Length:  26; Size in bytes:  344


The experiment runs on my environment requires 64 bytes(rather than 72 bytes in Code Fragment 5.1) but has the same increasing amount of 32 bytes whenever needed resize the list as the Code Fragment 5.1.

### R-5.2
In Code Fragment 5.1, we perform an experiment to compare the length of a Python list to its underlying memory usage. Determining the sequence of array sizes requires a manual inspection of the output of that program. Redesign the experiment so that the program outputs only those values of $k$ at which the existing capacity is exhausted. For example, on a system consistent with the results of Code Fragment 5.2, your program should output that the sequence of array capacities are 0, 4, 8, 16, 25, ....

In [3]:
import sys  # provides getsizeof function
data=[]
capacity = sys.getsizeof(data)
for k in range(1000):          # NOTE: must fix choice of n
    a = len(data)            # number of elements  
    b = sys.getsizeof(data)  # actual size in bytes
    if capacity != b:
        print(f'capacity: {a-1}')
#         print(f'Spend {(b - capacity) / 8} cyber dollers to resize the list')  # cyber dollers
        capacity = b

    data.append(None)        # increase length by one

capacity: 0
capacity: 4
capacity: 8
capacity: 16
capacity: 25
capacity: 35
capacity: 46
capacity: 58
capacity: 72
capacity: 88
capacity: 106
capacity: 126
capacity: 148
capacity: 173
capacity: 201
capacity: 233
capacity: 269
capacity: 309
capacity: 354
capacity: 405
capacity: 462
capacity: 526
capacity: 598
capacity: 679
capacity: 771
capacity: 874
capacity: 990


### R-5.3
Modify the experiment from Code Fragment 5.1 in order to demonstrate that Python’s list class occasionally shrinks the size of its underlying array when elements are popped from a list.

In [4]:
import sys  # provides getsizeof function


print('=== increasing the list ===')
data=[]
for k in range(27):          # NOTE: must fix choice of n
    a = len(data)            # number of elements  
    b = sys.getsizeof(data)  # actual size in bytes
    print('Length: {0:3d}; Size in bytes: {1:4d}'.format(a, b))
    data.append(None)        # increase length by one
    
print('\n=== shrinking the list ===')
while data:
    a = len(data)            # number of elements  
    b = sys.getsizeof(data)  # actual size in bytes
    print('Length: {0:3d}; Size in bytes: {1:4d}'.format(a, b))
    data.pop()        # increase length by one

=== increasing the list ===
Length:   0; Size in bytes:   64
Length:   1; Size in bytes:   96
Length:   2; Size in bytes:   96
Length:   3; Size in bytes:   96
Length:   4; Size in bytes:   96
Length:   5; Size in bytes:  128
Length:   6; Size in bytes:  128
Length:   7; Size in bytes:  128
Length:   8; Size in bytes:  128
Length:   9; Size in bytes:  192
Length:  10; Size in bytes:  192
Length:  11; Size in bytes:  192
Length:  12; Size in bytes:  192
Length:  13; Size in bytes:  192
Length:  14; Size in bytes:  192
Length:  15; Size in bytes:  192
Length:  16; Size in bytes:  192
Length:  17; Size in bytes:  264
Length:  18; Size in bytes:  264
Length:  19; Size in bytes:  264
Length:  20; Size in bytes:  264
Length:  21; Size in bytes:  264
Length:  22; Size in bytes:  264
Length:  23; Size in bytes:  264
Length:  24; Size in bytes:  264
Length:  25; Size in bytes:  264
Length:  26; Size in bytes:  344

=== shrinking the list ===
Length:  27; Size in bytes:  344
Length:  26; Size in

### R-5.4
Our DynamicArray class, as given in Code Fragment 5.3, does not support use of negative indices with `__getitem__`. Update that method to better match the semantics of a Python list.

In [5]:
def __getitem__(self, k):
    '''Return element at index k.
    '''
    if not -self._n <= k < self._n:
        raise IndexError('invalid index')
    return(self._A[k])

### R-5.5
Redo the justification of Proposition 5.1 assuming that the the cost of growing the array from size $k$ to size $2k$ is $3k$ cyber-dollars. How much should each append operation be charged to make the amortization work?

**Justification:** Let us assume that one cyber-dollar is enough to pay for the execution of each append operation in $S$, excluding the time spent for growing the array. Also, let us assume that growing the array from size $k$ to size $2k$ requires $3k$ cyber-dollars for the time spent initializing the new array. We shall charge each append operation **seven** cyber-dollars. Thus, we overcharge each append operation that does not cause an overflow by **six** cyber-dollars. Think of the **six** cyber-dollars profited in an insertion that does not grow the array as being “stored” with the cell in which the element was inserted. An overflow occurs when the array S has $2^i$ elements, for some integer $i ≥ 0$, and the size of the array used by the array representing $S$ is $2^i$. Thus, doubling the size of the array will require **$3 \cdot 2^i$** cyber-dollars. Fortunately, these cyber-dollars can be found stored in cells $2^{i−1}$ through $2^i − 1$. (See Figure 5.14.) Note that the previous overflow occurred when the number of elements became larger than $2^i−1$ for the first time, and thus the cyber-dollars stored in cells $2^{i−1}$ through $2^i − 1$ have not yet been spent. Therefore, we have a valid amortization scheme in which each operation is charged **seven** cyber-dollars and all the computing time is paid for. That is, we can pay for the execution of $n$ append operations using **$7n$** cyber-dollars. In other words, the amortized running time of each append operation is $O(1)$; hence, the total running time of $n$ append operations is $O(n)$.

<img style="center" src="R-5_5.jpeg" width="50%" height="50%">

### R-5.6
Our implementation of insert for the DynamicArray class, as given in Code Fragment 5.5, has the following inefficiency. In the case when a resize occurs, the resize operation takes time to copy all the elements from an old array to a new array, and then the subsequent loop in the body of insert shifts many of those elements. Give an improved implementation of the insert method, so that, in the case of a resize, the elements are shifted into their final position during that operation, thereby avoiding the subsequent shifting.

In [44]:
'''Code Fragment 5.3: An implementation of a DynamicArray class,
   using a raw array from the ctypes module as storage.
'''


import ctypes  # provides low-level arrays


class DynamicArray:
    '''A dynamic array class akin to a simplified Python list.
    '''
    def __init__(self):
        '''Create an empty array.
        '''
        self._n = 0                                  # count actual elements 
        self._capacity = 1                           # default array capacity 
        self._A = self._make_array(self._capacity)  # low-level array
        
    def __len__(self):
        '''Return number of elements stored in the array.
        '''
        return(self._n)
         
    def __getitem__(self, k):
        '''Return element at index k.
        '''
        if not 0 <= k < self._n:
            raise IndexError('invalid index')
        return self._A[k]                            # retrieve from array
    
    def append(self, obj):
        '''Add object to end of the array.
        '''
        if self._n == self._capacity:                # not enough room
            self._resize(2 * self._capacity)         # so double capacity
        self._A[self._n] = obj
        self._n += 1
        
    def _resize(self, c):                            # nonpublic utitity
        '''Resize internal array to capacity c.
        '''
        B = self._make_array(c)                      # new (bigger) array 
        for k in range(self._n):                     # for each existing value
            B[k] = self._A[k]
        self._A = B                                  # use the bigger array
        self._capacity = c
        
    def _make_array(self, c):                        # nonpublic utitity
        '''Return new array with capacity c.
        '''
        return(c * ctypes.py_object)()               # see ctypes documentation


    def insert(self, k, value):
        '''Code Fragment 5.5: Implementation of insert for our DynamicArray class.
           Insert value at index k, shifting subsequent values rightward.
        '''
         # (for simplicity, we assume 0 <= k <= n in this verion)
        if self._n == self._capacity:          # not enough room
            self._resize(2 * self._capacity)   # so double capacity
        for j in range(self._n, k, -1):        # shift rightmost first
            self._A[j] = self._A[j-1]
        self._A[k] = value                     # store newest element
        self._n += 1
        

class ImprovedInsertion(DynamicArray):
    def _resize(self, c, insert=False):               # nonpublic utitity
        '''The resize operation variation from Code Fragment 5.3 for R-5.6 solution
        '''
        B = self._make_array(c)                      # new (bigger) array 
        if insert:
            for k in range(self._n):                     # for each existing value
                B[k+1] = self._A[k]
        else:
            for k in range(self._n):                     # for each existing value
                B[k] = self._A[k]
        self._A = B                                  # use the bigger array
        self._capacity = c
        
    def insert(self, k, value):
        '''The insert operation variation from Code fregment 5.5 for R-5.6 solution
        '''
        # (for simplicity, we assume 0 <= k <= n in this verion)
        if self._n == self._capacity:
            print(f'Increasing capacity to: {self._capacity}')
            self._resize(2 * self._capacity, insert=True)
        else:
            for j in range(self._n, k, -1):        # shift rightmost first
                self._A[j] = self._A[j-1]
        self._A[k] = value
        self._n += 1

    
if __name__ == '__main__':
    import time
    t = time.time()
    da = DynamicArray()
    for i in range(10**2):
        da.insert(0, i)
    da.insert(0, 'a')
    print([i for i in da])
    print(time.time() - t)
    print()
    
    t = time.time()
    da2 = ImprovedInsertion()
    for i in range(10**2):
        da2.insert(0, i)
    da2.insert(0, 'a')
    print([i for i in da2])
    print(time.time() - t)

['a', 99, 98, 97, 96, 95, 94, 93, 92, 91, 90, 89, 88, 87, 86, 85, 84, 83, 82, 81, 80, 79, 78, 77, 76, 75, 74, 73, 72, 71, 70, 69, 68, 67, 66, 65, 64, 63, 62, 61, 60, 59, 58, 57, 56, 55, 54, 53, 52, 51, 50, 49, 48, 47, 46, 45, 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]
0.0027670860290527344

Increasing capacity to: 1
Increasing capacity to: 2
Increasing capacity to: 4
Increasing capacity to: 8
Increasing capacity to: 16
Increasing capacity to: 32
Increasing capacity to: 64
['a', 99, 98, 97, 96, 95, 94, 93, 92, 91, 90, 89, 88, 87, 86, 85, 84, 83, 82, 81, 80, 79, 78, 77, 76, 75, 74, 73, 72, 71, 70, 69, 68, 67, 66, 65, 64, 63, 62, 61, 60, 59, 58, 57, 56, 55, 54, 53, 52, 51, 50, 49, 48, 47, 46, 45, 44, 43, 42, 41, 40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14, 13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]


### R-5.7
Let A be an array of size $n ≥ 2$ containing integers from $1$ to $n − 1$, inclusive, with exactly one repeated. Describe a fast algorithm for finding the integer in $A$ that is repeated.

### Using Sorting as a Problem-Solving Tool
An even better algorithm for the ***element uniqueness problem*** is based on using sorting as a problem-solving tool. In this case, by sorting the sequence of elements, we are guaranteed that any duplicate elements will be placed next to each other. Thus, to determine if there are any duplicates, all we need to do is perform a sin- gle pass over the sorted sequence, looking for consecutive duplicates. A Python implementation of this algorithm is as follows:

```python
def unique2(S):
    '''Return True if there are no duplicate elements in sequence S.'''
    temp = sorted(S)                 # create a sorted copy of S
    for j in range(1, len(temp)):
        if S[j−1] == S[j]:
            return(False)            # found duplicate pair
    return(True)                     # if we reach this, elements were unique
```
    Code Fragment 3.8: Algorithm unique2 for testing element uniqueness.
The built-in function, `sorted`, as described in Section 1.5.2, produces a copy of the original list with elements in sorted order. It guarantees a worst-case running time of $O(nlogn)$; see Chapter 12 for a discussion of common sorting algorithms. Once the data is sorted, the subsequent loop runs in $O(n)$ time, and so the entire unique2 algorithm runs in $O(nlogn)$ time.

### R-5.8
Experimentally evaluate the efficiency of the `pop` method of Python’s list class when using varying indices as a parameter, as we did for `insert` on page 205. Report your results akin to Table 5.5.

In [11]:
import time


def efficiency_of_pop(N):
    print(f'==== N({N}) ====')
    # In the first case, we repeatedly pop with the default parameter(the last element of a list).
    data = [None for i in range(N)]
    t = time.time()
    while data:
        data.pop()
    print(time.time() - t)

    # In a second case, we repeatedly pop with the middle element of a list.
    data = list(range(N))
    t = time.time()
    while data:
        mid = len(data) // 2
        data.pop(mid)
    print(time.time() - t)

    # In a third case, we repeatedly pop with the first element of a list.
    data = list(range(N))
    t = time.time()
    while data:
        data.pop(0)
    print(time.time() - t)
    
if __name__ == '__main__':
    [efficiency_of_pop(10**n) for n in range(2, 7)]

==== N(100) ====
1.2159347534179688e-05
3.409385681152344e-05
3.814697265625e-05
==== N(1000) ====
0.00021004676818847656
0.0005278587341308594
0.0002779960632324219
==== N(10000) ====
0.0020151138305664062
0.011661052703857422
0.021746158599853516
==== N(100000) ====
0.008688926696777344
0.8558149337768555
1.8224198818206787
==== N(1000000) ====
0.10703182220458984
143.8716959953308
391.64639115333557


|         |          |         | $N$    |         |           |
|:--------|:---------|:--------|:-------|:--------|:----------|
|          | 100      | 1,000   | 10,000 | 100,000 | 1,000,000 |
|$k$ = $0$     | 1.22e-05 | 0.00021 | 0.0020 | 0.0087  | 0.11      |
|$k$ = $n$ // $2$| 3.41e-05 | 0.00053 | 0.12   | 0.8559  | 143.87    |
|$k$ = $n$     | 3.81e-05 | 0.00028 | 0.28   | 1.82    | 391.65    |

### R-5.9
Explain the changes that would have to be made to the program of Code Fragment 5.11 so that it could perform the Caesar cipher for messages that are written in an alphabet-based language other than English, such as Greek, Russian, or Hebrew.

To make Caesar cipher works for an alphabet-based language other than English we need:
1. We should figure out the whole characters of a language that we want to apply for cipher, so that we can map those characters to the respective numbers 0 to n(the total number characters of the language) through the `ord` method.
2. We need to change the fixed number of characters to a variable and the first character to another variable, as below:

In [18]:
'''Modified from Code Fragment 5.11 to support alphabet-based language other than English.
'''


class CaesarCipher:
    '''Class for doing encryption and decryption using a Caesar cipher.
    '''
    def __init__(self, shift, numbers_of_char, first_char):
        '''Construct Caesar cipher using given integer shift for rotation.
        '''
        encoder = [None] * numbers_of_char  # temp array for encryption
        decoder = [None] * numbers_of_char  # temp array for decryption
        for k in range(numbers_of_char):
            encoder[k] = chr((k + shift) % numbers_of_char + ord(first_char))
            decoder[k] = chr((k - shift) % numbers_of_char + ord(first_char))
        self._forward = ''.join(encoder)  # will store as string
        self._backward = ''.join(decoder)  # since fixed
        self._first_char = first_char
    
    def encrypt(self, message):
        '''Return string representing encripted message.
        '''
        return self._transform(message, self._forward)
    
    def decrypt(self, secret):
        '''Return decrypted message given encrypted secret.
        '''
        return self._transform(secret, self._backward)
    
    def _transform(self, original, code):
        '''Utility to perform transformation based on given code string.
        '''
        msg = list(original)
        for k in range(len(msg)):
            if msg[k].isupper():
                j = ord(msg[k]) - ord(self._first_char)  # index from 0 to 25 
                msg[k] = code[j]            # replace this character
                
        return(''.join(msg))
    
    
if __name__ == '__main__' :
    cipher = CaesarCipher(3, 26, 'A')
    message = "THE EAGLE IS IN PLAY; MEET AT JOE S."
    coded = cipher.encrypt(message)
    print('Secret: ', coded)
    answer = cipher.decrypt(coded)
    print('Message: ', answer)


Secret:  WKH HDJOH LV LQ SODB; PHHW DW MRH V.
Message:  THE EAGLE IS IN PLAY; MEET AT JOE S.


### R-5.10
The constructor for the CaesarCipher class in Code Fragment 5.11 can be implemented with a two-line body by building the forward and backward strings using a combination of the join method and an appropriate comprehension syntax. Give such an implementation.

In [19]:
def __init__(self, shift):
    self._forward = ''.join(chr((k + shift) % 26 + ord('A')) for k in range(26))
    self._backward = ''.join(chr((k - shift) % 26 + ord('A')) for k in range(26))

### R-5.11
Use standard control structures to compute the sum of all numbers in an $n × n$ data set, represented as a list of lists.

In [3]:
import random


def matrix_sum(S):
    result = 0
    for i in S:
        for j in i:
            result += j
    
    return(result)
        
        
N = 5
A = [[random.randint(0, 100) for i in range(N)] for j in range(N)]
print(A)
print(matrix_sum(A))

[[70, 9, 87, 83, 81], [61, 59, 41, 76, 48], [32, 77, 60, 47, 94], [27, 30, 25, 84, 35], [42, 88, 44, 9, 43]]
1352


### R-5.12
Describe how the built-in sum function can be combined with Python’s comprehension syntax to compute the sum of all numbers in an $n × n$ data set, represented as a list of lists.

In [6]:
N = 5
A = [[random.randint(0, 100) for i in range(N)] for j in range(N)]
print(A)
print(sum(sum(i) for i in A))

[[91, 52, 11, 27, 77], [1, 58, 4, 30, 88], [31, 84, 82, 92, 24], [8, 70, 26, 17, 69], [76, 25, 63, 81, 24]]
1211
